File: src/main/java/idv/kuma/Reels.java
package idv.kuma;

import java.util.ArrayList;
import java.util.List;

public final class Reels {

    private final List<Reel> reelList = new ArrayList<>();

    public Reels(List<List<String>> rawReels, RandomNumberGenerator randomNumberGenerator) {

        for (List<String> rawReel : rawReels) {
            Reel reel = new Reel(rawReel, randomNumberGenerator);
            reelList.add(reel);
        }


    }

    public void spin() {

        for (Reel reel : reelList) {
            reel.roll();
        }


    }


    public Screen getScreen() {

        List<List<String>> rawScreen = reelList.stream().map(
                reel -> reel.getScreenColumn(3)
        ).toList();

        return new Screen(rawScreen);
    }
}

File: javascript/src/Reels.ts
import {Screen} from "./Screen";
import {Reel} from "./Reel";
import {DesignatedNumberGenerator} from "./DesignatedNumberGenerator";
import {DBCTool} from "./DBCTool";

export class Reels {
    reels: Array<Reel>;

    private constructor(reels: Array<Array<string>>, randomNumberGenerator: DesignatedNumberGenerator) {
        this.reels = reels.map((reel: Array<string>): Reel => Reel.from(reel, randomNumberGenerator));
    }


    spin(): void {
        for (let i: number = 0; i < this.reels.length; ++i) {
            this.reels[i].spin();
        }
    }

    getScreen(): Screen {
        const rawScreen: Array<Array<string>> = [];
        for (let i: number = 0; i < this.reels.length; i++) {
            rawScreen.push(this.reels[i].getScreenColumn());
        }
        return Screen.from(rawScreen);
    }

    static create(randomNumberGenerator: DesignatedNumberGenerator, rawReels: Array<Array<string>>): Reels {
        return new Reels(rawReels, randomNumberGenerator);
    }

    getIndexes(): number[] {
        return this.reels.map((reel: Reel): number => reel.getIndex());
    }

    setIndexes(reelsIndexes: Array<number>) {
        DBCTool.require(() => reelsIndexes.length == this.reels.length, "Invalid reels indexes length");

        for (let i = 0; i < reelsIndexes.length; i++) {
            this.reels[i].setIndex(reelsIndexes[i]);
        }
    }
}